File: Problem_2/problem2.rs
// rustc problem2.rs

fn fib(n: i32) -> i32 {
    match n {
        1 | 2 => 1,
        3 => 2,
        _ => fib(n - 1) + fib(n - 2),
    }
}

fn main() {
    let mut num = 1;
    let mut result = 0;
    while fib(num) < 4_000_000 {
        if fib(num) % 2 == 0 {
            result += fib(num);
        }
        num += 1;
    }
    println!("{}", result);
}

File: 01-09/Problem_2/problem2.pl
%! [problem2].
%! problem2(X).
%! halt.

problem2(Answer) :- findall(N,fibEven(N),L), sumlist(L,Answer).

step(N,X) :- N = X; plus(3,N,N1), step(N1,X).

fibEven(X) :- step(2,N), fib(N,X), (X < 4000000 -> true; !, false).

fib(0, 1) :- !.
fib(1, 1) :- !.
fib(N, Result) :- N1 is N - 1, N2 is N - 2, fib(N1, Result1), fib(N2, Result2), Result is Result1 + Result2.
